nufi-official/metamask-key-tree
# Add bip32 ed25519 curve

**Motivation**
 - allow metamask extension (which uses this package), to derive keys consistent with Cardano

**Changes**
 - I tried to document everything in the code, but let me know if you are missing something
  - cardano derivation differs in two ways
    - root keypair is created from `entropy`, not `seed`
    - child keys are derived according https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf
  - in order to address these two differences, we had to add a new (what metamask considers) `curve` called `bip32Ed25519`
  - this curve implements `privateToPublicKey` and also helps the `bip39` differentiate that it should create the rootKeyPair from entropy, not seed (more on commented in the PR)
  - then we needed to add a new `deriver`. Deriver is an entity which exposes a single function `deriveChildKey` which  receives parent node and path. We called this deriver `cip3` as it describes exactly the changes we are making here https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/CIP-0003.md
  

**Notes**
 - only ~800 lines of code are actual changes, rest is test vectors json
 - already mentioned below, but I tried to make as little changes as possible to existing logic, so its easier for metamask to review this. Certain parts would definitely need a refactor but that is something I would leave for Metamask people do decide.
 
 **How to test**
  - there are test cases for everything, you can run `yarn test` locally
  - alternatively, currently the test cases use `kick find` mnemonic, so you can also check nufi if the keys are consistent

**TODO**
 - replace public key derivation so it does not use libsodium
 - replace `kick find` mnemonic test vectors with different one, ideally some official test vectors or vectors generated from well-used cardano lib
 - run checks (github actions command for it)
